plugins/cryptolib/cryptolib.cpp

```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
     }
 }
 
+//-----------------------------------------------------------------
+//  Simple cache for loaded keys
+//-----------------------------------------------------------------
+class CKeyCache
+{
+private:
+    CriticalSection csKeyCache;//guards serial modifications to the cache map
+    typedef std::unordered_map<string, Owned<CLoadedKey>> KeyCache;
+    KeyCache keyCache;
+
+public:
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)//
+    {
+        VStringBuffer searchKey("%s_%s_%s", isEmptyString(keyFS) ? "" : keyFS, isEmptyString(keyBuff) ? "" : keyBuff, isEmptyString(passphrase) ? "" : passphrase);
+        CriticalBlock block(csKeyCache);
+        KeyCache::iterator it = keyCache.find(searchKey.str());
+        CLoadedKey * ret = nullptr;
+        if (it != keyCache.end())//exists in cache?
+        {
+            ret = (*it).second;
+        }
+        else
+        {
+            CLoadedKey * pCLK = nullptr;
+            if (!isEmptyString(keyFS))
+            {
+                //Create CLoadedKey from filespec
+                if (isPublic)
+                    pCLK = loadPublicKeyFromFile(keyFS, passphrase);
```

Comment:
it seems we could load the keys outside of the critblock and we should only protect the lookup and the insert